logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,90 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageRepository
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.functional.onSuccess
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.datetime.Clock
+import kotlin.coroutines.CoroutineContext
+
+
+interface EphemeralMessageDeletionHandler {
+
+    fun startSelfDeletion(conversationId: ConversationId, messageId: String)
+    fun enqueuePendingSelfDeletionMessages()
+}
+
+
+// TODO:Mateusz: implement failure logic
+internal class EphemeralMessageDeletionHandlerImpl(
+    private val messageRepository: MessageRepository,
+    private val kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+    userSessionCoroutineScope: CoroutineScope
+) : EphemeralMessageDeletionHandler, CoroutineScope by userSessionCoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = kaliumDispatcher.default
+
+    private val onGoingSelfDeletionMessagesMutex = Mutex()
+
+    private val onGoingSelfDeletionMessages = mutableMapOf<Pair<ConversationId, String>, Unit>()
+    override fun startSelfDeletion(conversationId: ConversationId, messageId: String) {
+        launch {
+            onGoingSelfDeletionMessagesMutex.withLock {
+                val isSelfDeletionOutgoing = onGoingSelfDeletionMessages[conversationId to messageId] != null
+                if (isSelfDeletionOutgoing) return@launch
+
+                onGoingSelfDeletionMessages[conversationId to messageId] = Unit
+            }
+
+            messageRepository.getMessageById(conversationId, messageId).map { message ->
+                if (message is Message.Regular) enqueueSelfDeletion(message = message)
+            }
+        }
+    }
+
+    private suspend fun enqueueSelfDeletion(message: Message.Regular) {
+        if (message.isEphemeralMessage) {
+            val expirationData = message.expirationData!!
```

Comment:
same here, how is that better or requires this to change ?

Reply:
I generally try to avoid `!!` since it means instant crash if the value is not there. I guess in this case it can not happen since the `isEphemeralMessage = expirationData != null`.

But I actually have to go and look that up.

Why not replace `if (message.isEphemeralMessage) {}` message.expirationData?.let {} and let the compiler guarantee that the value is always there.

Reply:
in this case I agree that it does not change anything, but in some cases I like to make the language of what is going on more explicit, but in this case it indeed does not change any readibility